2021/classes/submarine.py

```diff
@@ -17,3 +18,30 @@ def deep_sonar_sweep(self):
             for r1, r2, r3 in zip(self.readings, self.readings[1:], self.readings[2:])
         ]
         return self._sonar_sweep(triple_step_sums)
+
+    def simple_position(self):
+        horizontal, depth = 0, 0
+
+        for direction, distance in self.move_commands:
```

Comment:
The biggest improvement I would suggest here is to introduce a more strongly modeled "move command".

Right now we're using a tuple which is _fine_, and it works, but generally I find that any time you are passing around a tuple or a dictionary, there is very likely another object hidden away that we could take advantage of.

Consider if a move command gained an additional vector, such that we now had something like `(direction, distance, time)`. In order to incorporate this change, we'd need to change every iterator over `move_commands`, as there's now an additional member to unpack.

Now, consider how we'd handle that same situation if we had a more strongly modelled move command.

```python
class MoveCommand:
    # I'm using type hints here
    def __init__(self, direction: str, distance: int):
        self.direction = direction
        self.distance = distance
```

Now instead of changing our iterators can look like this:
```python
for move_command in self.move_commands:
    ...
```

And when we add a new attribute (such as a `time`), our iterator doesn't need to change!

In a simple example like this, it's not a _huge_ deal, but in larger code bases, extracting out data into strongly modelled objects can often make it easier to make changes easier later on.

**TL;DR** - More objects is _almost always_ a good direction to go. It's very tough to have too many objects in a system. I've found that some of the nicest systems to work in are the ones that have many small objects that composed together make up the system as a whole.

Reply:
As a bit of a bonus, we could also extract `direction` into an `Enum`, as then we can operate on strongly typed definitions, rather than naked strings, which is always nice, as it has the same side effect of making change easier to push through the system later on.